README.md
# StatsD Rest Proxy

*Note: This repository is a combination of forks from https://github.com/InjectiveLabs/statsd-http-proxy and https://github.com/GoMetric/statsd-http-proxy. Along with an improved build story.*

StatsD uses UDP connections, and  can not be used directly from browser. This server is a HTTP proxy to StatsD, useful for sending metrics to StatsD from frontend by AJAX.

Requests may be optionally authenticated using JWT tokens.

## Table of contents

* [Installation](#installation)
* [Nginx config](#nginx-config)
* [Usage](#usage)
* [Client Interactions](#client-interactions)

## Installation

```
# clone repo
make
```

## Nginx config

Configuration of Nginx balancer:

```config
server {
    listen 443 http2;
    server_name statsd-proxy.example.com;
    ssl on;
    ssl_certificate     /etc/pki/nginx/ssl.crt;
    ssl_certificate_key /etc/pki/nginx/ssl.key;
    upstream statsd_proxy {
        keepalive 100;
        server statsd-proxy-1:8825 max_fails=0;
        server statsd-proxy-2:8825 max_fails=0;
    }

    location / {
        proxy_pass http://statsd_proxy;
        proxy_redirect off;
        proxy_http_version 1.1;
        proxy_set_header Connection "keep-alive";
    }
}
```

## Caddy config

```config
localhost {
    handle /statsd* {
        uri strip_prefix /statsd
        # suppress external access to /metrics and /heartbeat endpoints
        redir /metrics / permanent
        redir /heartbeat / permanent
        reverse_proxy statsd-proxy:80
    }
}
```

## Usage

* Run server (HTTP):

```bash
statsd-http-proxy \
    --verbose \
    --http-host=127.0.0.1 \
    --http-port=8080 \
    --statsd-host=127.0.0.1 \
    --statsd-port=8125 \
    --jwt-secret=somesecret \
    --metric-prefix=prefix.subprefix
```

* Run server (HTTPS):

```bash
statsd-http-proxy \
    --verbose \
    --http-host=127.0.0.1 \
    --http-port=433 \
    --tls-cert=cert.pem \
    --tls-key=key.pem \
    --statsd-host=127.0.0.1 \
    --statsd-port=8125 \
    --jwt-secret=somesecret \
    --metric-prefix=prefix.subprefix
```

Print server version and exit:

```bash
statsd-http-proxy --version
```

Command line arguments:

| Parameter       | Description                          | Default value                                                                                |
|--------------------|--------------------------------------|----------------------------------------------------------------------------------------------|
| verbose            | Print debug info to stderr           | Optional. Default false                                                                      |
| http-host          | Host of HTTP server                  | Optional. Default 127.0.0.1. To accept connections on any interface, set to ""               |
| http-port          | Port of HTTP server                  | Optional. Default 8825                                                                       |
| http-timeout-read  | The maximum duration in seconds for reading the entire request, including the body | Optional. Defaults to 2 seconds                |
| http-timeout-write | The maximum duration in seconds before timing out writes of the respons | Optional. Defaults to 2 seconds                           |
| http-timeout-idle  | The maximum amount of time in seconds to wait for the next request when keep-alives are enabled | Optional. Defaults to 5 seconds   |
| tls-cert           | TLS certificate for the HTTPS        | Optional. Default "" to use HTTP. If both tls-cert and tls-key set, HTTPS is used            |
| tls-key            | TLS private key for the HTTPS        | Optional. Default "" to use HTTP. If both tls-cert and tls-key set, HTTPS is used            |
| statsd-host        | Host of StatsD instance              | Optional. Default 127.0.0.1                                                                  |
| statsd-port        | Port of StatsD instance              | Optional. Default 8125                                                                       |
| jwt-secret         | JWT token secret                     | Optional. If not set, server accepts all connections                                         |
| metric-prefix      | Prefix, added to any metric name     | Optional. If not set, do not add prefix                                                      |
| version            | Print version of server and exit     | Optional                                                                                     |
| prometheus-compat  | Enforce the prometheus data model on all incoming metrics, meaning some characters will be filtered/changed | Optional              |
| normalize          | All metrics will be converted to lowercase strings | Optional                                                                       |

## Client Interactions

Sample code to send metric in browser with JWT token in header:

```javascript
$.ajax({
    url: 'http://127.0.0.1:8080/count/',
    method: 'POST',
    headers: {
        'X-JWT-Token': 'some-jwt-token'
        'Content-Type': 'application/json'
    },
    data: {
        metric: 'some.key.name',
        value: 100500
    }
});
```
## Supported metrics

For the general reference see https://www.librato.com/docs/kb/collect/collection_agents/stastd/#

All metrics accept `tags` as comma-separated key=value pairs (InfluxDB tag format):

```javascript
data: {
    metric: 'some.key.name',
    value: 100500,
    tags: 'env=prod,locale=en-us'
}
```

### `count`

Adds count to the bucket. Expected `value` as integer. By default `value` is 0.

### `gauge`

Sets the gauge metric. Expected `value` as integer. Before setting negative gauge, it needs to be set to `0`.

### `timing`

Adds timing to the bucket. Expected `value` as milliseconds integer. Default is `0`.

### `set`

Adds value in a set bucket. Sets are a relatively new concept in recent versions of StatsD. Sets track the number of unique elements belonging to a group. At each flush interval, the statsd backend will push the number of unique elements in the set as a single gauge value.

### Batch Writes

Sample code to send a group metrics in browser with JWT token in header:

```javascript
$.ajax({
    url: 'http://127.0.0.1:8080/batch/',
    method: 'POST',
    headers: {
        'X-JWT-Token': 'some-jwt-token'
        'Content-Type': 'application/json'
    },
    data: [
      {
        metric: 'some.key.name',
        value: 100500,
        tags: 'env=prod,locale=en-us',
        metric_type: 'count'
      },
      {
        metric: 'some.other.key.name',
        value: 1,
        metric_type: 'gauge'
      }
    ]
});
```

This allows us to send multiple metrics at once, saving significant bandwidth/processing time for larger groups of writes.

## Legacy Pattern

You can also send metrics in using the legacy pattern:

```javascript
$.ajax({
    url: 'http://127.0.0.1:8080/count/some.key.name',
    method: 'POST',
    headers: {
        'X-JWT-Token': 'some-jwt-token'
        'Content-Type': 'application/json'
    },
    data: {
        value: 100500
    }
});
```

We define the metric name in the URL rather than the data object. This _does_ slightly restrict the allowed characters a metric name can have (because allowed URL characters are more limited).

All other suggestions/requirements from above apply to metrics generated this way.


CHANGELOG.md
## 2.0.3
  * improve internal metrics some
    * Track dropped tags and processing queue length
  * increase internal queue depth
    * For higher through-put situations, 10k vs 1k queue depth may matter.

## 2.0.2
  * handle bad metric type gracefully

## 2.0.1
  * Simplify passing variables in processor object
  * track dropped metrics during http request, too

## 2.0
  * Channel for processing metrics
    * significantly improve client performance (since client doesn't have to wait for entire processing pipelien to finish)
  * move batch writes to separate endpoint
    * allows for individual stats in a batch to be different types

## 1.7
  * batch writes into the `/:type` endpoint
    * many metrics in one write to the proxy, vastly improving performance over the wire
  * Go 1.19

## 1.6
  * prometheus-compat filtering for incoming metrics
  * fix metric prefix settings

## 1.5.2
  * bug with empty tag list fixed

## 1.5.1
  * allow for metric names in URLs _or_ in body

## 1.5
  * middleware to get better prometheus stat generation (RED/etc.)
  * Golang 1.18.5

## 1.4
  * metric names in message body
  * `/metrics` endpoint for prometheus-style stats tracking of proxy

## 1.3.1
  * Vendor updates

## 1.3
  * Go 1.18

## 1.2
  * Bump vendoring for jwt (fixes bugs in implementation)
  * merged changes from multiple projects

## 1.1
  * pull vendoring into local repo
  * build against Go 1.17
  * better linting/testing in build
  * pull in gometric statsd client
  * make JSON input values consistent (no `n` vs. `value` vs. `dur`) so clients can more easily send data

## 1.0-stable
  * use `alexcesaro/statsd`
  * adapt the stats interface

## 1.0-dev
  * Default HTTP port changed to 8825
  * Configure HTTP timeouts through `--http-timeout-read`, `--http-timeout-write` and `--http-timeout-idle`
  * Minumum required Go version is 1.8

## 0.9 (2019-01-29)
  * TLS Secure connection listening

## 0.8 (2018-02-24)
  * Added support of preflight CORS OPTIONS requests with header `X-JWT-Token`
  * Added support of authentication with passing token in query string instead of header `X-JWT-Token`

## 0.7 (2017-12-25)
  * Binary renamed to `statsd-http-proxy`


main.go
package main

import (
	"flag"
	"fmt"
	"net/http"
	_ "net/http/pprof"
	"os"
	"runtime"
	"strings"
	"time"

	vmmetrics "github.com/VictoriaMetrics/metrics"
	"github.com/civic-eagle/statsd-http-proxy/proxy"
	"github.com/civic-eagle/statsd-http-proxy/proxy/process"
	"github.com/civic-eagle/statsd-http-proxy/proxy/statsdclient"
	log "github.com/sirupsen/logrus"
)

// Version is a current git tag
// Injected by compilation flag
var Version = "Unknown"

// BuildDate is a date of build
// Injected by compilation flag
var BuildTime = "Unknown"

// BuildUser is the user that built
// Injected by compilation flag
var BuildUser = "Unknown"

// HTTP connection params
const defaultHTTPHost = "127.0.0.1"
const defaultHTTPPort = 8825
const defaultHTTPReadTimeout = 2
const defaultHTTPWriteTimeout = 2
const defaultHTTPIdleTimeout = 5

// StatsD connection params
const defaultStatsDHost = "127.0.0.1"
const defaultStatsDPort = 8125

func main() {
	startTime := time.Now()
	// track start time metric in the correct place
	_ = vmmetrics.NewGauge("app_uptime_secs_total",
		func() float64 {
			return float64(time.Since(startTime).Seconds())
		})

	// logging configuration
	log.SetFormatter(&log.JSONFormatter{})
	log.SetOutput(os.Stdout)

	// declare command line options
	var httpHost = flag.String("http-host", defaultHTTPHost, "HTTP listening address")
	var httpPort = flag.Int("http-port", defaultHTTPPort, "HTTP Port")
	var httpReadTimeout = flag.Int("http-timeout-read", defaultHTTPReadTimeout, "The maximum duration in seconds for reading the entire request, including the body")
	var httpWriteTimeout = flag.Int("http-timeout-write", defaultHTTPWriteTimeout, "The maximum duration in seconds before timing out writes of the response")
	var httpIdleTimeout = flag.Int("http-timeout-idle", defaultHTTPIdleTimeout, "The maximum amount of time in seconds to wait for the next request when keep-alives are enabled")
	var tlsCert = flag.String("tls-cert", "", "TLS certificate to enable HTTPS")
	var tlsKey = flag.String("tls-key", "", "TLS private key  to enable HTTPS")
	var statsdHost = flag.String("statsd-host", defaultStatsDHost, "StatsD listening address")
	var statsdPort = flag.Int("statsd-port", defaultStatsDPort, "StatsD Port")
	var metricPrefix = flag.String("metric-prefix", "", "Prefix of metric name")
	var tokenSecret = flag.String("jwt-secret", "", "Secret to encrypt JWT")
	var verbose = flag.Bool("verbose", false, "Verbose")
	var promFilter = flag.Bool("prometheus-compat", false, "Enforce prometheus data model compatibility on incoming metrics")
	var normalize = flag.Bool("normalize", false, "Ensure all metrics (and tags) are lower case strings")
	var version = flag.Bool("version", false, "Show version")
	var profilerHTTPort = flag.Int("profiler-http-port", 0, "Start profiler localhost")

	// get flags
	flag.Parse()

	if *verbose {
		log.SetLevel(log.DebugLevel)
	} else {
		log.SetLevel(log.InfoLevel)
	}

	// show version and exit
	if *version {
		log.WithFields(log.Fields{"Version": Version, "BuildTime": BuildTime, "BuildUser": BuildUser}).Info("Version Info")
		os.Exit(0)
	}

	// start profiler
	if *profilerHTTPort > 0 {
		// enable block profiling
		runtime.SetBlockProfileRate(1)

		// start debug server
		profilerHTTPAddress := fmt.Sprintf("localhost:%d", *profilerHTTPort)
		go func() {
			log.WithFields(log.Fields{"Address": profilerHTTPAddress}).Info("Profiler started")
			log.Info(fmt.Sprintf("Open 'http://%s/debug/pprof/' in you browser or use 'go tool pprof http://%s/debug/pprof/heap' from console", profilerHTTPAddress, profilerHTTPAddress))
			log.Info("See details about pprof in https://golang.org/pkg/net/http/pprof/")
			log.Info(http.ListenAndServe(profilerHTTPAddress, nil))
		}()
	}

	/*
		Create processor and related tooling
		Because this is async, we create it outside the
		http server
	*/
	// prepare metric prefix
	if *metricPrefix != "" && (*metricPrefix)[len(*metricPrefix)-1:] != "_" {
		*metricPrefix = *metricPrefix + "_"
	}
	if *normalize {
		*metricPrefix = strings.ToLower(*metricPrefix)
	}

	// create StatsD Client
	statsdClient := statsdclient.NewGoMetricClient(*statsdHost, *statsdPort)
	// open StatsD connection
	statsdClient.Open()
	defer statsdClient.Close()

	// build processor
	processor := processor.NewProcessor(
		statsdClient,
		*metricPrefix,
		*promFilter,
		*normalize,
	)

	/*
		Multiple processing threads so we don't get bottle-necked on one processor
		Since each individual object on the channel is unique, we don't need
		state between processor threads!
	*/
	for thread := 1; thread <= 4; thread++ {
		go processor.Process()
	}

	// start proxy server
	proxyServer := proxy.NewServer(
		*httpHost,
		*httpPort,
		*httpReadTimeout,
		*httpWriteTimeout,
		*httpIdleTimeout,
		*tlsCert,
		*tlsKey,
		*tokenSecret,
		*verbose,
	)

	proxyServer.Listen()
}
